File: tests/test-path-bar.c
/*
 * Copyright 2010 Intel Corporation.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms and conditions of the GNU Lesser General Public License,
 * version 2.1, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for
 * more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St - Fifth Floor, Boston, MA 02110-1301 USA.
 * Boston, MA 02111-1307, USA.
 *
 */

#include <mx/mx.h>

static void
add_cb (MxPathBar *bar)
{
  gchar *string = g_strdup_printf ("Crumb %d", mx_path_bar_get_level (bar) + 1);
  mx_path_bar_push (bar, string);
  g_free (string);
}

static void
remove_cb (MxPathBar *bar)
{
  mx_path_bar_pop (bar);
}

static void
toggle_editable_cb (MxPathBar *bar)
{
  ClutterActor *parent = clutter_actor_get_parent (CLUTTER_ACTOR (bar));
  gboolean editable = !mx_path_bar_get_editable (bar);
  mx_path_bar_set_editable (bar, editable);
  if (MX_IS_BIN (parent))
    mx_bin_set_fill (MX_BIN (parent), editable, FALSE);
  else if (MX_IS_BOX_LAYOUT (parent))
    clutter_container_child_set (CLUTTER_CONTAINER (parent),
                                 CLUTTER_ACTOR (bar),
                                 "x-fill", editable,
                                 NULL);
}

static void
relabel_cb (MxPathBar *bar)
{
  const gchar *string = mx_path_bar_get_text (bar);
  mx_path_bar_set_label (bar, 1, string ? string : "");
}

int
main (int argc, char **argv)
{
  MxWindow *window;
  MxApplication *app;
  ClutterActor *stage, *bar, *button, *hbox;

  app = mx_application_new (&argc, &argv, "Test PathBar", 0);

  window = mx_application_create_window (app);
  stage = (ClutterActor *)mx_window_get_clutter_stage (window);

  bar = mx_path_bar_new ();
  mx_path_bar_set_clear_on_change (MX_PATH_BAR (bar), TRUE);

  hbox = mx_box_layout_new ();

  button = mx_button_new_with_label ("Add crumb");
  g_signal_connect_swapped (button, "clicked",
                            G_CALLBACK (add_cb), bar);
  clutter_container_add_actor (CLUTTER_CONTAINER (hbox), button);

  button = mx_button_new_with_label ("Remove crumb");
  g_signal_connect_swapped (button, "clicked",
                            G_CALLBACK (remove_cb), bar);
  clutter_container_add_actor (CLUTTER_CONTAINER (hbox), button);

  button = mx_button_new_with_label ("Toggle editable");
  g_signal_connect_swapped (button, "clicked",
                            G_CALLBACK (toggle_editable_cb), bar);
  clutter_container_add_actor (CLUTTER_CONTAINER (hbox), button);

  button = mx_button_new_with_label ("Re-label first crumb");
  g_signal_connect_swapped (button, "clicked",
                            G_CALLBACK (relabel_cb), bar);
  clutter_container_add_actor (CLUTTER_CONTAINER (hbox), button);

  /* -a for 'alternative packing'... */
  if (argc > 1 && g_str_equal (argv[1], "-a"))
    {
      ClutterActor *vbox = mx_box_layout_new ();
      mx_box_layout_set_orientation (MX_BOX_LAYOUT (vbox), MX_ORIENTATION_VERTICAL);
      clutter_container_add_actor (CLUTTER_CONTAINER (vbox), bar);
      clutter_container_add_actor (CLUTTER_CONTAINER (vbox), hbox);
      mx_window_set_child (window, vbox);
      mx_path_bar_set_editable (MX_PATH_BAR (bar), TRUE);
    }
  else
    {
      MxToolbar *toolbar = mx_window_get_toolbar (window);
      clutter_container_add_actor (CLUTTER_CONTAINER (toolbar), bar);
      mx_window_set_child (window, hbox);
    }

  clutter_actor_show (stage);

  mx_application_run (app);

  return 0;
}

File: tests/javascript/test-path-bar.js
#!/usr/bin/env gjs
/*
 * Copyright 2010 Intel Corporation.
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms and conditions of the GNU Lesser General Public License,
 * version 2.1, as published by the Free Software Foundation.
 *
 * This program is distributed in the hope it will be useful, but WITHOUT ANY
 * WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for
 * more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin St - Fifth Floor, Boston, MA 02110-1301 USA.
 * Boston, MA 02111-1307, USA.
 *
 */

const Clutter = imports.gi.Clutter;
const Mx = imports.gi.Mx;

Clutter.init (null);

let stage = Clutter.Stage.get_default ();
stage.title = "Test Path Bar"
let focus_man = Mx.FocusManager.get_for_stage (stage);

let vbox = new Mx.BoxLayout ({orientation: Mx.Orientation.VERTICAL});
stage.add_actor (vbox);
vbox.set_width (stage.width);

let bar = new Mx.PathBar ({"clear-on-change": true});
vbox.add_actor (bar, -1);
vbox.child_set_property (bar, "x-align", 0);
vbox.child_set_property (bar, "x-fill", false);

let hbox = new Mx.BoxLayout ();
vbox.add_actor (hbox, -1);

let push_button = new Mx.Button ( {label: "Add crumb"} )
hbox.add_actor (push_button, -1);

let pop_button = new Mx.Button ( {label: "Remove crumb"} );
hbox.add_actor (pop_button, -1);

let editable_button = new Mx.Button ( {label: "Toggle editable"} );
hbox.add_actor (editable_button, -1);

let rename_button = new Mx.Button ( {label: "Re-label 1st button"} );
hbox.add_actor (rename_button, -1);

push_button.connect ("clicked",
                     function (b) {
                       bar.push ("Crumb" + (bar.level + 1));
                     });

pop_button.connect ("clicked",
                    function (b) {
                      bar.pop ();
                    });

editable_button.connect ("clicked",
                         function (b) {
                           bar.editable = !bar.editable;
                           vbox.child_set_property (bar, "expand", bar.editable);
                           vbox.child_set_property (bar, "x-fill", bar.editable);
                         });

rename_button.connect ("clicked",
                       function (b) {
                         if (bar.editable)
                           bar.set_label (1, bar.entry.text);
                       });

stage.show ();
Clutter.main ();

stage.destroy ();
